rero_ils/modules/documents/serializers/ris.py

```diff
@@ -0,0 +1,171 @@
+from rero_ils.modules.documents.serializers.base import \
+    BaseDocumentFormatterMixin
+from rero_ils.modules.documents.utils import create_contributions
+from rero_ils.utils import get_i18n_supported_languages
+
+
+class RISSerializer:
+    """BibTeX serializer for records."""
+
+    def serialize(self, pid, record, links_factory=None, **kwargs):
+        """Serialize a single record and persistent identifier.
+
+        :param pid: Persistent identifier instance.
+        :param record: Record instance.
+        :param links_factory: Factory function for record links.
+        """
+        Document.post_process(record)
+        record = record.replace_refs()
+        contributions = create_contributions(
+            record.get('contribution', [])
+        )
+        if contributions:
+            record['contribution'] = contributions
+        return RISFormatter(record=record).format()
+
+    def serialize_search(self, pid_fetcher, search_result, links=None,
+                         item_links_factory=None, **kwargs):
+        """Serialize a search result.
+
+        :param pid_fetcher: Persistent identifier fetcher.
+        :param search_result: Elasticsearch search result.
+        :param links: Dictionary of links to add to response.
+        """
+        records = [
+            RISFormatter(record=hit['_source']).format()
+            for hit in search_result['hits']['hits']
+        ]
+
+        return stream_with_context(records)
+
+
+class RISFormatter(BaseDocumentFormatterMixin):
+    """RIS formatter class."""
+
+    def __init__(self, record, doctype_mapping=None):
+        """Initialize RIS formatter with the specific record."""
+        super().__init__(record)
+        config = current_app.config.get('RERO_ILS_EXPORT_MAPPER').get('ris')
+        language = request.args.get("lang", current_i18n.language)
+        if not language or language not in get_i18n_supported_languages():
+            language = current_app.config.get('BABEL_DEFAULT_LANGUAGE', 'en')
+        self._language = language
+        self._doctype_mapping = doctype_mapping \
+            or config.get('doctype_mapping')
+
+    def format(self):
+        """Return RIS export for single record."""
+        return self._format()
+
+    def _format(self):
+        """Format record to RIS."""
+        export_fields = [
+            'TY', 'ID', 'TI', 'T2', 'AU', 'A2', 'DA', 'SP', 'EP', 'CY',
+            'LA', 'PB', 'SN', 'UR', 'KW', 'ET', 'DO', 'VL', 'IS', 'PP',
+            'Y1', 'PY']
+        out = ''
+        out += self._fetch_fields(export_fields)
+        out += 'ER -\n'
+        return out
+
+    def _doctype_mapper(self, main_type: str, sub_type: str = None):
+        """Document type mapper.
+
+        :param: main_type: main document type.
+        :param: sub_type: subtype of main document type.
+        :return: mapped RIS reference type.
+        """
+        for k, v in self._doctype_mapping.items():
```

Comment:
maybe `k` and `v` name are not very relevant here.

Reply:
you are right